core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java

```diff
@@ -498,32 +498,31 @@ public void delete(K key, InternalCacheEntry<K, V> value, RemovalCause cause) {
                listener.onEntryChosenForEviction(key, value);
             }
          }
-      });
+      }).removalListener(listener);
    }
 
    static <K, V> Caffeine<K, V> caffeineBuilder() {
       return (Caffeine<K, V>) Caffeine.newBuilder();
    }
 
    /**
-    * Handles eviction logic. That is that it will order notifications, passivation and notify listeners as applicable
-    * @param key key of the evicted entry
-    * @param value evicted entry
+    * Handles the logic of what occurs when an entry is evicted. That is it will acquire the orderer for the key if necessary
+    * (not null), and then passivate the entry and subsequently notify the listeners, all in a non blocking fashion.
+    * Note this method specifically handles when concurrent activations or writes occur to the same key.
+    * @param entry evicted entry
     * @param orderer used to guarantee ordering between other operations. May be null when an operation is already ordered
     * @param passivator Passivates the entry to the store if necessary
     * @param evictionManager Handles additional eviction logic. May be null if eviction is also not required
     * @param <K> key type of the entry
     * @param <V> value type of the entry
     * @return stage that when complete all of the eviction logic is complete
     */
-   public static <K, V> CompletionStage<Void> handleEviction(K key, InternalCacheEntry<K, V> value,
-         NonBlockingOrderer orderer, PassivationManager passivator, EvictionManager<K, V> evictionManager,
-         DataContainer<K, V> dataContainer) {
+   public static <K, V> CompletionStage<Void> handleEviction(InternalCacheEntry<K, V> entry, NonBlockingOrderer orderer,
+         PassivationManager passivator, EvictionManager<K, V> evictionManager, DataContainer<K, V> dataContainer,
+         CompletionStage<Void> writeDelay) {
```

Comment:
Is `writeDelay` the best name for this as it's a future not a duration. How about something like `evictionSyncFuture`?

Reply:
Well technically it is used solely for the purpose of delaying the operation and we have called similar things this in the async interceptor stack.